File: client/go/WebSocketUpstreamTransport.go
package modcdp

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"sync"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

type WebSocketUpstreamTransport struct {
	UpstreamTransport
	URL     string
	Conn    net.Conn
	writeMu sync.Mutex
	ctx     context.Context
	cancel  context.CancelFunc
}

func NewWebSocketUpstreamTransport(url ...string) *WebSocketUpstreamTransport {
	transport := &WebSocketUpstreamTransport{}
	if len(url) > 0 {
		transport.URL = url[0]
	}
	return transport
}

func (t *WebSocketUpstreamTransport) Update(config map[string]any) {
	if config == nil {
		return
	}
	for _, key := range []string{"ws_url", "cdp_url", "url"} {
		if value, ok := config[key].(string); ok && value != "" {
			t.URL = value
			return
		}
	}
}

func (t *WebSocketUpstreamTransport) GetServerConfig() map[string]any {
	if t.URL == "" {
		return map[string]any{}
	}
	return map[string]any{"loopback_cdp_url": t.URL}
}

func (t *WebSocketUpstreamTransport) Connect() error {
	if t.URL == "" {
		return fmt.Errorf("upstream.mode=ws requires upstream.ws_url or launcher-provided ws_url")
	}
	resolvedURL, err := websocketURLFor(t.URL)
	if err != nil {
		return err
	}
	t.URL = resolvedURL
	t.ctx, t.cancel = context.WithCancel(context.Background())
	conn, _, _, err := ws.Dial(t.ctx, t.URL)
	if err != nil {
		return err
	}
	t.Conn = conn
	return nil
}

func (t *WebSocketUpstreamTransport) Send(message map[string]any) error {
	body, err := json.Marshal(message)
	if err != nil {
		return err
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	return wsutil.WriteClientText(t.Conn, body)
}

func (t *WebSocketUpstreamTransport) Close() error {
	if t.cancel != nil {
		t.cancel()
	}
	if t.Conn != nil {
		return t.Conn.Close()
	}
	return nil
}

File: client/python/modcdp/WebSocketUpstreamTransport.py
from __future__ import annotations

import json
from urllib.request import urlopen
from typing import Any

from websocket import create_connection

from .UpstreamTransport import UpstreamTransport


class WebSocketUpstreamTransport(UpstreamTransport):
    mode = "ws"
    endpoint_kind = "raw_cdp"

    def __init__(self, url: str | None = None, timeout_s: float = 10) -> None:
        super().__init__()
        self.url = url or ""
        self.timeout_s = timeout_s
        self.ws: Any | None = None

    def update(self, config: dict[str, Any] | None = None) -> "WebSocketUpstreamTransport":
        config = config or {}
        url = config.get("ws_url") or config.get("cdp_url") or config.get("url")
        if url:
            self.url = str(url)
        return self

    def getServerConfig(self) -> dict[str, Any]:
        return {"loopback_cdp_url": self.url} if self.url else {}

    def connect(self) -> None:
        if not self.url:
            raise RuntimeError("upstream.mode='ws' requires upstream.ws_url or launcher-provided ws_url.")
        self.url = _websocket_url_for(self.url)
        self.ws = create_connection(self.url, timeout=self.timeout_s)

    def send(self, message: dict[str, Any]) -> None:
        if self.ws is None:
            raise RuntimeError("CDP websocket is not connected.")
        self.ws.send(json.dumps(message))

    def recv(self) -> Any:
        if self.ws is None:
            raise RuntimeError("CDP websocket is not connected.")
        return self.ws.recv()

    def close(self) -> None:
        if self.ws is not None:
            self.ws.close()
        self.ws = None


def _websocket_url_for(endpoint: str) -> str:
    if endpoint.startswith(("ws://", "wss://")):
        return endpoint
    http_endpoint = endpoint if "://" in endpoint else f"http://{endpoint}"
    with urlopen(http_endpoint.rstrip("/") + "/json/version", timeout=10) as response:
        version = json.loads(response.read().decode())
    ws_url = version.get("webSocketDebuggerUrl")
    if not isinstance(ws_url, str) or not ws_url:
        raise RuntimeError("upstream.ws_url HTTP discovery returned no webSocketDebuggerUrl")
    return ws_url
